lib/src/rewrite.rs

```diff
@@ -514,12 +517,18 @@ pub fn move_commits(
         connected_target_commits_internal_parents.insert(commit.id().clone(), new_parents);
     }
 
-    // Compute the roots of `target_commits`.
-    let target_roots: HashSet<_> = connected_target_commits_internal_parents
-        .iter()
-        .filter(|(commit_id, parents)| target_commit_ids.contains(commit_id) && parents.is_empty())
-        .map(|(commit_id, _)| commit_id.clone())
-        .collect();
+    // Compute the roots of `target_commits` if not provided.
+    let target_roots: HashSet<_> = if target_roots.is_empty() {
```

Comment:
nit: make `target_roots` an `Option` instead so it's consistent (but useless, iiuc) with an empty list

Reply:
Alternatively, `target_commits` and `target_roots` could be enum `Revisions(_)|Roots(_)`? I assume some connectivity checks can be skipped if `Roots(_)`.

Reply:
> Alternatively, `target_commits` and `target_roots` could be enum `Revisions(_)|Roots(_)`? I assume some connectivity checks can be skipped if `Roots(_)`.

That's definitely an option. I was thinking that since this was a library function, we might want to allow customizing consumers to specify the target set of commits and the new roots of the set anyway. WDYT? Otherwise we can also change this to an enum like you suggested.

Reply:
I don't think library function should support weird setup such as roots out of the target commits. If we can optimize for common configuration or guard against invalid parameters, I would choose restricted enum than free parameters. (They can be separate function entry points if that makes sense.)

Reply:
I'm wondering if it's fine to revisit this tweak in a separate PR? I think it would be easier for me to follow up after these commits are merged, then have to individually go back and edit each commit to update the API.

Reply:
Sorry, I didn't mean the optimization for `Root(..)` should be added in this PR. I just meant the argument can be changed to `Option` or `Revisions(_)|Roots(_)`.